### PR TITLE
[FLINK-4084] Add configDir parameter to CliFrontend and flink shell script

### PR DESCRIPTION
**Pull request by alkagin**
Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.
- [ X] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)
- [X ] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added
- [X ] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

Modified flink shell script and CliFrontend to provide an option `configDir` to specify a custom configuration folder.

### CODE REVIEW

File: docs/apis/cli.md [hunk 1]
```diff
@@ -187,6 +187,8 @@ Action "run" compiles and runs a program.
                                           java.net.URLClassLoader}.
      -d,--detached                        If present, runs the job in detached
                                           mode
+     -D,--configDir </path/to/confdir>    The configuration directory with which
```
Comment (mxm) [inline docs/apis/cli.md]: `-D`? I think that can be easily confused. We don't have to have a short option.

File: flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java [hunk 2]
```diff
@@ -132,9 +132,9 @@
 
 
 
-	private final Configuration config;
+	private Configuration config;
```
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java]: Why did you remove the `final` flag here?

File: flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java [hunk 3]
```diff
@@ -132,9 +132,9 @@
...
 
-	private final FiniteDuration clientTimeout;
+	private FiniteDuration clientTimeout;
```
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java]: Same as above.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java [hunk 4]
```diff
@@ -76,6 +76,9 @@
 	static final Option SAVEPOINT_DISPOSE_OPTION = new Option("d", "dispose", true,
 			"Disposes an existing savepoint.");
 
+	static final Option CONFIGDIR_OPTION = new Option("D", "configDir", true,
```
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java]: I would be for removing the short option because it can easily be confused with `-d`.

File: flink-clients/src/test/java/org/apache/flink/client/CliFrontendRunTest.java [hunk 5]
```diff
@@ -98,6 +98,12 @@ public void testRun() {
 				RunOptions options = CliFrontendParser.parseRunCommand(parameters);
 				assertEquals("expectedSavepointPath", options.getSavepointPath());
 			}
+			// test configure configDir
+			{
+				String[] parameters = {"-D", "expectedConfigDirectory", getTestJarPath()};
```
Comment (mxm) [inline flink-clients/src/test/java/org/apache/flink/client/CliFrontendRunTest.java]: Both long and short option should be tested (though I think we can remove the short option).

File: flink-dist/src/main/flink-bin/bin/flink [hunk 6]
```diff
@@ -17,6 +17,20 @@
...
+function abspath() {
+    if [ -d "$1" ]; then
+        # dir
+        (cd "$1"; pwd)
+    elif [ -f "$1" ]; then
+        # file
+        if [[ $1 == */* ]]; then
+            echo "$(cd "${1%/*}"; pwd)/${1##*/}"
+        else
+            echo "$(pwd)/$1"
```
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: Have you looked in `config.sh`? There is a function for resolving symbolic links. What happens in case of cyclic symbolic links?

File: flink-dist/src/main/flink-bin/bin/flink [hunk 7]
```diff
@@ -33,6 +47,22 @@ while [ -L "$target" ]; do
...
 done
 
+
+#Search --configDir into the parameters and set it as FLINK_CONF_DIR
+args=("$@")
+for i in "${!args[@]}"; do
+    if [[ ${args[$i]} = "--configDir" || ${args[$i]} = "-D" ]]; then
+        dir=${args[$(($i+1))]}
+        if [ -d "$dir" ]; then
+            FLINK_CONF_DIR= abspath ${dir}
```
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: Should never assign to the variable because of the space after `=`.

**Comment by mxm** [comment]
Thanks for the pull request! I've made some comments. The pull requests also needs to be rebased against the latest master.

File: docs/apis/cli.md (hunk 1 above)
Comment (alkagin) [inline docs/apis/cli.md]: I totally agree, I've added a short option 'cause I think it was required. I will remove asap

File: flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java (hunk 2 above)
Comment (alkagin) [inline flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java]: I removed final from clientTimeout and config because I need to reassign them in [CliFrontend.java#L219](https://github.com/alkagin/flink/blob/bb9203923edeb02258455712db579d09dd3e7e62/flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java#L219), as far as I understand `--configDir` option must be available just in `run` mode. Is it correct? Otherwise I can read it in the main and instantiate correctly CliFrontend passing the configDir.

File: flink-dist/src/main/flink-bin/bin/flink [hunk 8]
```diff
@@ -17,20 +17,41 @@
...
+
+    echo "$bar"
+}
+
+target=$(followSymLink "$0")
+
+#Search --configDir into the parameters and set it as FLINK_CONF_DIR
+args=("$@")
+for i in "${!args[@]}"; do
+    if [[ ${args[$i]} = "--configDir" || ${args[$i]} = "-D" ]]; then
```
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: We still have the old `-D` parameter here.

File: flink-dist/src/main/flink-bin/bin/flink [hunk 9]
```diff
@@ -17,20 +17,41 @@
...
+        dir=$(followSymLink "${args[$(($i+1))]}" )
+        if [ -d "$dir" ]; then
+            FLINK_CONF_DIR=`cd "${dir}"; pwd -P`
```
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: Can we make this `FLINK_CONF_DIR=$(cd "${dir}" && pwd -P)`?

File: flink-clients/src/test/java/org/apache/flink/client/CliFrontendRunTest.java [hunk 10]
```diff
@@ -100,15 +100,22 @@ public void testRun() {
...
 			// test jar arguments
+            {
+                String[] parameters =
+                        {"-m", "localhost:6123", getTestJarPath(), "-arg1", "value1", "justavalue", "--arg2", "value2"};
+                RunOptions options = CliFrontendParser.parseRunCommand(parameters);
+                assertEquals("-arg1", options.getProgramArgs()[0]);
+                assertEquals("value1", options.getProgramArgs()[1]);
+                assertEquals("justavalue", options.getProgramArgs()[2]);
+                assertEquals("--arg2", options.getProgramArgs()[3]);
+                assertEquals("value2", options.getProgramArgs()[4]);
```
Comment (mxm) [inline flink-clients/src/test/java/org/apache/flink/client/CliFrontendRunTest.java]: Tabs have been converted to spaces here. That's why the diff looks odd. Please use tabs for indention.

File: flink-dist/src/main/flink-bin/bin/flink [hunk 11]
```diff
@@ -17,20 +17,41 @@
...
 # Note: we can't use the readlink utility here if we want to be POSIX
 # compatible.
-iteration=0
-while [ -L "$target" ]; do
-    if [ "$iteration" -gt 100 ]; then
-        echo "Cannot resolve path: You have a cyclic symlink in $target."
+followSymLink() {
+    local iteration=0
+    local bar=$1
+    while [ -L "$bar" ]; do
```
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: +1 for local variables but please don't call them `bar` 😄 How about `local target=$1`?

File: flink-dist/src/main/flink-bin/bin/flink [hunk 12]
```diff
@@ -17,20 +17,41 @@
...
+
+    echo "$bar"
+}
+
+target=$(followSymLink "$0")
+
+#Search --configDir into the parameters and set it as FLINK_CONF_DIR
+args=("$@")
+for i in "${!args[@]}"; do
+    if [ ${args[$i]} = "--configDir" ]; then
```
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: I think this needs another space after `]`.

File: flink-dist/src/main/flink-bin/bin/flink [hunk 13]
```diff
@@ -17,20 +17,41 @@
...
+        dir=$(followSymLink "${args[$(($i+1))]}" )
```
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: Please add a space after `$(` or remove the space after `}"`.

File: flink-dist/src/main/flink-bin/bin/flink [hunk 14]
```diff
@@ -17,20 +17,41 @@
...
+        if [ -d "$dir" ]; then
+            FLINK_CONF_DIR=`cd "${dir}"; pwd -P`
```
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: Could we replace this with `FLINK_CONF_DIR=$(cd "${dir}" && pwd -P)`?

File: flink-clients/src/test/java/org/apache/flink/client/CliFrontendRunTest.java (hunk 10 above)
Comment (alkagin) [inline flink-clients/src/test/java/org/apache/flink/client/CliFrontendRunTest.java]: Damn intellij upgrade, it should reset config style.

File: flink-dist/src/main/flink-bin/bin/flink (hunk 14 above)
Comment (alkagin) [inline flink-dist/src/main/flink-bin/bin/flink]: I found a similar command [here](https://github.com/alkagin/flink/blob/76283d3f0fbf5454b208ba1887d847acea09640d/flink-dist/src/main/flink-bin/bin/config.sh#L132).

File: flink-dist/src/main/flink-bin/bin/flink (hunk 12 above)
Comment (alkagin) [inline flink-dist/src/main/flink-bin/bin/flink]: There is already one, we need two?

File: flink-dist/src/main/flink-bin/bin/flink (hunk 14 above)
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: I suppose it doesn't matter much but I like the chained version better.

Comment (alkagin) [inline flink-dist/src/main/flink-bin/bin/flink]: Added chained version, the second command runs only if the first was successful; i think it is a more correct behaviour.

File: flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java (hunk 2 above)
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java]: Could we make the `configDir` parameter a general parameter that goes before `run`, `list`, `cancel`, etc.? Don't know if that is too tricky but it would make sense because it is a general configuration key.

`flink --configDir /my/custom/dir run -p 10 /path/to/jar`

That way the config parameter wouldn't have to be listed for every action. Would make the help look less verbose.

You could parse the config parameter in the constructor and initialize the config only once. Then continue with the normal parsing.

Comment (alkagin) [inline flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java]: > Could we make the configDir parameter a general parameter that goes before run, list, cancel, etc.?
Sounds good. `configDir` must be positional after `flink` or I may do something like this `flink run --configDir /my/custom/dir -p 10 /path/to/jar`?

> You could parse the config parameter in the constructor and initialize the config only once. Then continue with the normal parsing.
Now `configDir` is traversal on all actions, i can do it.

Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/CliFrontend.java]: Let's make it positional to prevent confusion.

File: flink-dist/src/main/flink-bin/bin/flink (hunk 12 above)
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: You're right, it's not necessary.

**Comment by mxm** [comment]
Thanks for the update. We're getting there :) Great work so far!

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java [hunk 15]
```diff
@@ -459,4 +476,14 @@ public static InfoOptions parseInfoCommand(String[] args) throws CliArgsExceptio
 		}
 	}
 
+	public static MainOptions parseMainCommand(String[] args) throws CliArgsException {
+		try {
+			DefaultParser parser = new DefaultParser();
+			CommandLine line = parser.parse(MAIN_OPTIONS, args, false);
```
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java]: You're parsing the entire arguments here. What if the user jar has `--configDir` as parameter?

File: flink-clients/src/test/java/org/apache/flink/client/CliFrontendMainTest.java [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+
+	@Test
+	public void testMain() throws CliArgsException, FileNotFoundException, MalformedURLException {
+			// test configure configDir
+			{
+				String[] parameters = {"--configDir", "expectedConfigDirectory", getTestJarPath()};
+				MainOptions options = CliFrontendParser.parseMainCommand(parameters);
+				assertEquals("expectedConfigDirectory", options.getConfigDir());
+			}
+	}
```
Comment (mxm) [inline flink-clients/src/test/java/org/apache/flink/client/CliFrontendMainTest.java]: We need more tests here to assure the parameter is correctly parsed.

File: flink-clients/src/test/java/org/apache/flink/client/CliFrontendMainTest.java [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+	public static void init() {
+		CliFrontendTestUtils.pipeSystemOutToNull();
+		CliFrontendTestUtils.clearGlobalConfiguration();
+	}
+
+	@Test
+	public void testMain() throws CliArgsException, FileNotFoundException, MalformedURLException {
+			// test configure configDir
+			{
+				String[] parameters = {"--configDir", "expectedConfigDirectory", getTestJarPath()};
```
Comment (mxm) [inline flink-clients/src/test/java/org/apache/flink/client/CliFrontendMainTest.java]: This is not a valid invocation of the CliFrontend.

Comment (mxm) [inline flink-clients/src/test/java/org/apache/flink/client/CliFrontendMainTest.java]: i.e. there is no "action" before the jar.

**Comment by mxm** [comment]
@alkagin I've been on vacations. Getting back to you soon!

File: docs/apis/cli.md [hunk 18]
```diff
@@ -187,6 +187,8 @@ Action "run" compiles and runs a program.
                                           java.net.URLClassLoader}.
      -d,--detached                        If present, runs the job in detached
                                           mode
+        --configDir </path/to/confdir>    The configuration directory with which
```
Comment (mxm) [inline docs/apis/cli.md]: I think there is a tab, we use only spaces for formatting here.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java [hunk 19]
```diff
@@ -451,4 +479,25 @@ public static InfoOptions parseInfoCommand(String[] args) throws CliArgsExceptio
 		}
 	}
 
+	public static MainOptions parseMainCommand(String[] args) throws CliArgsException {
+
+		// drop all arguments after an action
+		final List<String> params= Arrays.asList(args);
```
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java]: space missing

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java [hunk 20]
```diff
@@ -451,4 +479,25 @@ public static InfoOptions parseInfoCommand(String[] args) throws CliArgsExceptio
...
+			int index = params.indexOf(action);
+			if(index != -1) {
+				args = Arrays.copyOfRange(args, 0, index);
+				break;
+			}
+		}
+
+		try {
+			DefaultParser parser = new DefaultParser();
+			CommandLine line = parser.parse(MAIN_OPTIONS, args, false);
```
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java]: Actually you can use `CommandLine line = parser.parse(MAIN_OPTIONS, args, true);` to halt once you reach a non-arg.

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java [hunk 21]
```diff
@@ -451,4 +479,25 @@ public static InfoOptions parseInfoCommand(String[] args) throws CliArgsExceptio
...
 	}
 
+	public static MainOptions parseMainCommand(String[] args) throws CliArgsException {
+
+		// drop all arguments after an action
+		final List<String> params= Arrays.asList(args);
+		for (String action: CliFrontend.ACTIONS) {
+			int index = params.indexOf(action);
+			if(index != -1) {
+				args = Arrays.copyOfRange(args, 0, index);
```
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java]: I think dropping the args is not necessary if you use the parsing below.

File: flink-clients/src/main/java/org/apache/flink/client/cli/ProgramOptions.java [hunk 22]
```diff
@@ -148,4 +148,5 @@ public boolean getDetachedMode() {
 	public String getSavepointPath() {
 		return savepointPath;
 	}
+
```
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/cli/ProgramOptions.java]: The changes in this file are not necessary. Could you revert?

File: flink-clients/src/test/java/org/apache/flink/client/CliFrontendMainTest.java [hunk 23]
```diff
@@ -0,0 +1,39 @@
...
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+import java.io.FileNotFoundException;
+import java.net.MalformedURLException;
+
+import static org.apache.flink.client.CliFrontendTestUtils.getTestJarPath;
+import static org.junit.Assert.assertEquals;
+
+public class CliFrontendMainTest {
```
Comment (mxm) [inline flink-clients/src/test/java/org/apache/flink/client/CliFrontendMainTest.java]: Maybe `CliFrontendMainArgsTest`?

File: flink-clients/src/test/java/org/apache/flink/client/CliFrontendRunTest.java [hunk 24]
```diff
@@ -110,6 +110,8 @@ public void testRun() {
 				assertEquals("--arg2", options.getProgramArgs()[3]);
 				assertEquals("value2", options.getProgramArgs()[4]);
 			}
+
+
```
Comment (mxm) [inline flink-clients/src/test/java/org/apache/flink/client/CliFrontendRunTest.java]: Not necessary change.

File: flink-dist/src/main/flink-bin/bin/flink [hunk 25]
```diff
@@ -17,21 +17,39 @@
 # limitations under the License.
```
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: You change the mode from 644 to 755.

File: flink-dist/src/main/flink-bin/bin/flink [hunk 26]
```diff
@@ -17,21 +17,39 @@
...
+        iteration=$((iteration + 1))
+    done
+
+    echo "$target"
+}
+
+target=$(followSymLink "$0")
+
+#Check if --configDir is present and set is value as FLINK_CONF_DIR
+if [ "$1" = "--configDir" ]; then
```
Comment (mxm) [inline flink-dist/src/main/flink-bin/bin/flink]: Here you assume that `--configDir` is the first parameter but later on you parse for it until you find the action command. Why do you set both the environment variable and pass it on as a parameter?

**Comment by alkagin** [comment]
I am on vacations, I will address your comments as soon as I get back.

File: docs/apis/cli.md (hunk 18 above)
Comment (chobeat) [inline docs/apis/cli.md]: This is a space, I checked. Do you want it to be aligned to the beginning of the other lines or is it ok like it is?

File: flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java [hunk 27]
```diff
@@ -241,8 +255,11 @@ private static Options getSavepointOptionsWithoutDeprecatedOptions(Options optio
 	 * Prints the help for the client.
 	 */
 	public static void printHelp() {
-		System.out.println("./flink <ACTION> [OPTIONS] [ARGUMENTS]");
+		System.out.println("./flink [CONFIGDIR] <ACTION> [ACTION-OPTIONS] [ARGUMENTS]");
```
Comment (mxm) [inline flink-clients/src/main/java/org/apache/flink/client/cli/CliFrontendParser.java]: Shouldn't this be `./flink [--configDir <configDir>] <ACTION> [ACTION-OPTIONS] [ARGUMENTS]`?

File: docs/apis/cli.md (hunk 18 above)
Comment (mxm) [inline docs/apis/cli.md]: Ah ok. Looks good. Thanks!

Comment (mxm) [inline docs/apis/cli.md]: Actually, it should be aligned and `configDir` is not part of the run options.
